In [1]:
# Include main project path in the system path for python
import sys
import os
import re
import pandas as pd
import urllib
sys.path.insert( 0,os.path.abspath(os.path.join(os.getcwd(), '..')))

# Import the breadboard module
from breadboard import BreadboardClient

## Get parameters for images

In [10]:
# initialize the client
client = BreadboardClient(config_path='API_CONFIG.json')

In [11]:
imagenames = ['10-09-2018_00_27_44_TopA','10-09-2018_00_27_44_TopB']
params = ['FB_field_13_V','ImagFreq0', 'cylinder hold time', 'unixtime']
params = "list_bound_only"
df = client.get_images_df(imagenames, params)

In [12]:
df.head()

imagename             x      unixtime
0  10-09-2018_00_27_44_TopA  1.539045e+09  1.539045e+09
1  10-09-2018_00_27_44_TopB  1.539045e+09  1.539045e+09

In [5]:
# note the API does a lot of caching so the data might be a little outdated

## Update notes for a single image

In [6]:
params_update  = {
            'notes': "test"
        }

In [7]:
resp = client.update_image( df.at[0,'id'],df.at[0,'imagename'], params_update)

In [8]:
resp.status_code

200

## Random stuff

In [4]:
# initialize the client
client = BreadboardClient(config_path='API_CONFIG_2.json')

In [5]:
df = client.get_images_df(['2019-02-27_20-28-25_SensicamQE','2019-02-27_20-32-42_SensicamQE'], paramsin = "list_bound_only", xvar='reWhirrFreq')
df.head()

imagename      x  holdTime  TOF  evapEndTOPMHz  \
0  2019-02-27_20-28-25_SensicamQE  4.725      10.0  0.0         1777.0   
1  2019-02-27_20-32-42_SensicamQE  4.675      10.0  0.0         1777.0   

       unixtime  reWhirrFreq  accordionHoldSpacing  
0  1.551300e+09        4.725                   4.4  
1  1.551299e+09        4.675                   4.4

In [6]:
resp = client.get_images_json(pd.read_clipboard(header=None)[0].tolist())

EmptyDataError: No columns to parse from file

In [18]:
resp

<Response [414]>

Note: the 414 error means the query string is too long. Instead of getting a long list, we need to make a post request.

Can translate long queries into date ranges, but that doesn't avoid bad shots or specific subsets

In [7]:
df = client.get_images_df_clipboard(xvar='numberOfSpins')

In [8]:
df.head()

imagename     x  whirrFreq  TOPside2p5Quad  \
0  2019-02-03_20-57-32_SensicamQE  50.0       4.75          -3.525   
1  2019-02-03_20-55-48_SensicamQE  50.0       4.75          -3.525   
2  2019-02-03_20-54-50_SensicamQE  50.0       4.75          -3.525   
3  2019-02-03_20-50-53_SensicamQE  60.0       4.75          -3.525   
4  2019-02-03_20-50-20_SensicamQE  55.0       4.75          -3.525   

   numberOfSpins      unixtime  totalRampTime  rampFraction  evapEndTOPMHz  
0           50.0  1.549227e+09           0.25          0.95         1777.1  
1           50.0  1.549227e+09           0.25          0.95         1777.1  
2           50.0  1.549227e+09           0.25          0.95         1777.1  
3           60.0  1.549227e+09           0.25          0.95         1777.1  
4           55.0  1.549227e+09           0.25          0.95         1777.1

In [15]:
df

imagename     x      unixtime  numberOfSpins  \
0   2019-02-04_03-07-18_SensicamQE  12.0  1.549250e+09           12.0   
1   2019-02-04_03-06-46_SensicamQE   2.0  1.549250e+09            2.0   
2   2019-02-04_03-06-14_SensicamQE  15.0  1.549250e+09           15.0   
3   2019-02-04_03-05-42_SensicamQE  26.0  1.549250e+09           26.0   
4   2019-02-04_03-05-10_SensicamQE   7.0  1.549250e+09            7.0   
5   2019-02-04_03-04-37_SensicamQE  47.0  1.549249e+09           47.0   
6   2019-02-04_03-04-05_SensicamQE  20.0  1.549249e+09           20.0   
7   2019-02-04_03-03-33_SensicamQE   3.0  1.549249e+09            3.0   
8   2019-02-04_03-03-01_SensicamQE  20.0  1.549249e+09           20.0   
9   2019-02-04_03-02-28_SensicamQE  10.0  1.549249e+09           10.0   
10  2019-02-04_03-01-57_SensicamQE  14.0  1.549249e+09           14.0   
11  2019-02-04_03-01-25_SensicamQE  53.0  1.549249e+09           53.0   
12  2019-02-04_03-00-52_SensicamQE  13.0  1.549249e+09           13.0   
13  2019-02-04_03-00-20_SensicamQE  32.0  1.549249e+09           32.0   
14  2019-02-04_02-59-48_SensicamQE  18.0  1.549249e+09           18.0   
15  2019-02-04_02-59-16_SensicamQE  50.0  1.549249e+09           50.0   
16  2019-02-04_02-58-43_SensicamQE  35.0  1.549249e+09           35.0   
17  2019-02-04_02-58-12_SensicamQE  56.0  1.549249e+09           56.0   
18  2019-02-04_02-57-39_SensicamQE  23.0  1.549249e+09           23.0   
19  2019-02-04_02-57-06_SensicamQE  44.0  1.549249e+09           44.0   
20  2019-02-04_02-56-34_SensicamQE  19.0  1.549249e+09           19.0   
21  2019-02-04_02-56-02_SensicamQE  11.0  1.549249e+09           11.0   
22  2019-02-04_02-55-29_SensicamQE   5.0  1.549249e+09            5.0   
23  2019-02-04_02-54-57_SensicamQE   9.0  1.549249e+09            9.0   
24  2019-02-04_02-54-25_SensicamQE  16.0  1.549249e+09           16.0   
25  2019-02-04_02-53-53_SensicamQE  17.0  1.549249e+09           17.0   
26  2019-02-04_02-53-21_SensicamQE   0.0  1.549249e+09            0.0   
27  2019-02-04_02-52-50_SensicamQE  38.0  1.549249e+09           38.0   
28  2019-02-04_02-52-18_SensicamQE   6.0  1.549249e+09            6.0   
29  2019-02-04_02-51-45_SensicamQE  41.0  1.549249e+09           41.0   
30  2019-02-04_02-51-13_SensicamQE  29.0  1.549249e+09           29.0   
31  2019-02-04_02-50-40_SensicamQE   4.0  1.549249e+09            4.0   
32  2019-02-04_02-50-08_SensicamQE   1.0  1.549249e+09            1.0   
33  2019-02-04_02-49-36_SensicamQE  59.0  1.549249e+09           59.0   
34  2019-02-04_02-49-03_SensicamQE   8.0  1.549249e+09            8.0   

    totalRampTime  whirrFreq  rampFraction  evapEndTOPMHz  TOPside2p5Quad  
0            0.25       4.75           0.7         1777.0          -3.525  
1            0.25       4.75           0.7         1777.0          -3.525  
2            0.25       4.75           0.7         1777.0          -3.525  
3            0.25       4.75           0.7         1777.0          -3.525  
4            0.25       4.75           0.7         1777.0          -3.525  
5            0.25       4.75           0.7         1777.0          -3.525  
6            0.25       4.75           0.7         1777.0          -3.525  
7            0.25       4.75           0.7         1777.0          -3.525  
8            0.25       4.75           0.7         1777.0          -3.525  
9            0.25       4.75           0.7         1777.0          -3.525  
10           0.25       4.75           0.7         1777.0          -3.525  
11           0.25       4.75           0.7         1777.0          -3.525  
12           0.25       4.75           0.7         1777.0          -3.525  
13           0.25       4.75           0.7         1777.0          -3.525  
14           0.25       4.75           0.7         1777.0          -3.525  
15           0.25       4.75           0.7         1777.0          -3.525  
16           0.25       4.75           0.7         1777.0          -3.525  
17           0.25       4.75

In [10]:
import datetime
import re

def timestr_to_datetime(time_string):
    time_string = re.sub(' ','0',time_string[0:19])
    return datetime.datetime.strptime(time_string,'%m-%d-%Y_%H_%M_%S')

def isoformat_datetime(dt_object):
    return dt_object.isoformat()+'Z'

In [11]:
timestr = '10-09-2018_00_21_57_TopA'
dt_object = timestr_to_datetime(timestr)

In [12]:
'2018-10-09T00:21:57Z' == isoformat_datetime(dt_object)

True

In [13]:
type(dt_object)==datetime.datetime

True

In [14]:
type('44')==str

True

In [15]:
TIMEFORMATS = {
    'BEC1': 'blah',
    'FERMI3':'poo'
}

In [16]:
TIMEFORMATS['BEC1']

'blah'

In [7]:
from dateutil import parser

In [8]:
parser.parse('2019-02-27_20-39-28')

ValueError: ('Unknown string format:', '2019-02-27_20-39-28')

In [13]:
resp = client.get_images_json(imagenames)

In [16]:
jsonresp = resp.json()

In [24]:
jsonresp[0]['run']['runtime']

'2018-10-09T00:22:00Z'

In [27]:
import dateutil

In [35]:
int(dateutil.parser.parse('2018-10-09T00:22:00Z').timestamp())

1539044520

In [39]:
pd.__version__

'0.22.0'

In [38]:
import pandas as pd